### ISSUE TITLE
Incompatible with Yarn 2 PnP linker

### ISSUE THREAD

**Issue by rdicroce** [eclipse/wildwebdeveloper#651]
This is probably an issue with the TypeScript language server rather than WWD itself, but I'm not 100% sure.

I set up a project with Yarn 2, which uses the PnP linker by default, meaning that there is no node_modules directory. When I opened a TS file, I got errors about being unable to find modules that I'm importing from, even though the necessary dependencies are listed in package.json. I tried a few different things and finally "fixed" the issue by configuring Yarn to use the node-modules linker instead, so now there is a node_modules directory.

Weirdly, everything worked fine under the PnP linker for a .js file I was looking at. I'm not sure if the difference is .js vs .ts, or if it's because the .js used require() whereas .ts used import.

**Comment by mickaelistria** [comment]
Indeed, this seems like an issue in the language server as Wild Web Developer itself does no resolution nor parsing. Can you please report this issue (ideally with a minimal example) to https://github.com/theia-ide/typescript-language-server ?

**Comment by jabby** [comment]
There is an issue about PnP on typescript-language-server repository. See https://github.com/theia-ide/typescript-language-server/issues/193

**Comment by rdicroce** [comment]
@jabby Thanks, looks like that issue was opened right after I opened this one. The reporter of that issue appears to be using emacs and is having the same issue, so that reaffirms that this is a problem with the language server and not WWD itself.

**Comment by rdicroce** [comment]
@vrubezhny Apologies for the very late reply. I just tried this with the latest release and it still does not work. After some investigation, I think I've figured out why, and WWD is the cause this time.

First, we have to look at what typescript-language-server did to implement PnP compatibility. Their answer was to lean on the tsserver wrapper generated by `@yarnpkg/sdks`. So the wrapper needs to be generated, and WWD doesn't do that, but that's only a mild inconvenience since the user could do that manually.

However, generating the wrapper is not enough, because we need WWD to actually use that wrapper, and that's where WWD gets into trouble. The wrapper is generated in the project directory, under `.yarn/sdks/`, and typescript-language-server would ordinarily look there to find the tsserver to use. But WWD passes the `--tsserver-path` option, which overrides the search logic, and so the wrapper is not called.

If WWD omitted that option, then the wrapper should get called and that would get a lot of stuff to work, but probably still not everything. In particular, since PnP keeps the dependencies in a zip file, WWD would have to understand the zip file URLs that are returned for features like "go to definition" - see [here](https://yarnpkg.com/getting-started/editor-sdks).

**Comment by mickaelistria** [comment]
@rdicroce seems like you got a good track to follow. Would you be willing to submit a PR for that?

**Comment by mickaelistria** [comment]
Wild Web Developer doesn't assume any package.json exists, it assumes as little as possible and can work with plain .js files without need for package.json nor any other file. Yes, it's indeed the idea that `--tsserver-path` is a fallback, but since Wild Web Developer starts only 1 instance of the language server, I don't think it's possible to give flexibility here according to the projects being developed and the files being opened.

**Comment by rdicroce** [comment]
According to plugin.xml, the JSTS and Angular language servers are not declared as singletons. So if I'm reading [LSP4E's code](https://github.com/eclipse/lsp4e/blob/607e86e697e49e4f1cb1969ecdf9009cbfa47df4/org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java#L538) correctly, it should already be spawning a separate server for each project, since typescript-language-server does not say it supports workspaces. Which is exactly what we need here. We just also need typescript-language-server to treat `--tsserver-path` as a fallback instead of an override, and it should work.

The Angular language server may have a similar problem here with `--tsProbeLocations` but I'm not sure about that.

**Comment by mickaelistria** [comment]
@rdicroce OK, indeed it's 1 LS per project at the moment, I forgot about that. So you're probably right. Can you please open the issue to typescript-language-server about making `--tsserver-path` a fallback?

**Comment by rdicroce** [comment]
Actually it turns out that typescript-language-server will already attempt to use `require.resolve` to find typescript as a last resort. So removing `--tsserver-path` makes it work both with and without package.json. I'll open a PR to do that shortly.

As I suspected, that doesn't fix everything:
- The "open definition" feature is indeed broken, so I think I'll have to open a separate PR in LSP4E to fix that.
- The Angular language server also doesn't work for the same reason typescript-language-server didn't: WWD isn't passing the wrapped typescript with `--tsProbeLocations`. This one is a bit more problematic because we need to know the project rootUri when `start()` is called, and LSP4E currently doesn't provide any way to achieve that. So that will also need a separate issue in LSP4E.
